rust-engine/src/interning.rs

```diff
@@ -0,0 +1,287 @@
+        function(input)
+    }
+}
+
+impl<T: Hash + Eq + Clone + Send> InterningTable<T> {
+    fn try_intern(&mut self, value: &T) -> Option<Interned<T>> {
+        Some(if let Some(interned) = self.ids.get(value) {
+            *interned
+        } else {
+            let index = self.items.len();
+            self.items.push(value.clone());
+            let handle = Interned {
+                phantom: PhantomData,
+                index: index.try_into().ok()?,
+            };
+            self.ids.insert(value.clone(), handle);
+            handle
+        })
+    }
+    fn get(&self, interned: Interned<T>) -> &T {
+        &self.items[interned.index as usize]
+    }
+
+    /// Creates a global `LazyLock` interning table prepopulated with `values`,
+    /// and returns both the lock and the corresponding `Interned<T>` handles.
+    ///
+    /// # Panics
+    ///
+    /// Panics if `values` contains duplicates (by `Eq`) or if `N` is greater
+    /// than `u32::MAX`.
+    pub const fn new_with_values<const N: usize>(
+        values: fn() -> [T; N],
+    ) -> (LazyLockNewWithValue<T, N>, [Interned<T>; N]) {
+        assert!(N < u32::MAX as usize);
+        let mut i = 0;
+        let mut interned_values: [Interned<T>; N] = [Interned {
+            phantom: PhantomData,
+            index: 0,
+        }; N];
+        while i < N {
+            interned_values[i].index = i as u32;
+            i += 1;
+        }
+        let lazy_lock = LazyLock::new(ExplicitClosure(values, |values| {
+            let values = values();
+            {
+                // Ensure `value` has no duplicate.
+                let set: HashSet<_> = values.iter().collect();
+                if set.len() != values.len() {
+                    panic!("new_with_values: the input has duplicates");
+                }
+            }
+
+            let mut table = InterningTable::default();
+            for value in values {
+                if table.try_intern(&value).is_none() {
+                    unreachable!(
+                        "we asserted `N < u32::MAX`, the length of the internal vector `table` should be less than `u32::MAX`"
+                    )
+                }
+            }
+            Mutex::new(table)
+        }));
+        (lazy_lock, interned_values)
+    }
+}
+
+/// A type alias representing a lazily initialized `Mutex<InterningTable<T>>`
+/// backed by a fixed-size array initializer.
+///
+/// This is the return type of [`InterningTable::new_with_values`].
+pub type LazyLockNewWithValue<T, const N: usize> =
+    LazyLock<Mutex<InterningTable<T>>, ExplicitClosure<fn() -> [T; N], Mutex<InterningTable<T>>>>;
+
+/// Types that have a single, process‑global interning table.
+///
+/// Implement this for your type to opt in to interning:
+/// provide a `static` (usually a `LazyLock<Mutex<InterningTable<Self>>>`)
+/// and return a reference to it.
+pub trait Internable: Sized + Hash + Eq + Clone + Send + Any + 'static {
```

Comment:
Why the Any bound here? I am guessing for `type_name`, but I don't think that needs it https://doc.rust-lang.org/std/any/fn.type_name.html

Reply:
Oh, you're right, thanks! I always thought it needed the Any bound!
Will remove it.